From 5079ea209b4fc02ef0f16278fb4596fa48a53b13 Mon Sep 17 00:00:00 2001
From: Scott Atchley <atchleyes@ornl.gov>
Date: Wed, 29 Jun 2016 15:01:55 -0400
Subject: [PATCH] sm: set the device's pid when not using config

Fix suggested by John Jenkins <jenkins@mcs.anl.gov>

sm: fix leaks when closing a conn
---
 src/plugins/ctp/sm/ctp_sm_api.c | 29 +++++++++++++++++++++++------
 1 file changed, 23 insertions(+), 6 deletions(-)

diff --git a/src/plugins/ctp/sm/ctp_sm_api.c b/src/plugins/ctp/sm/ctp_sm_api.c
index 91ce369..88ff44d 100644
--- a/src/plugins/ctp/sm/ctp_sm_api.c
+++ b/src/plugins/ctp/sm/ctp_sm_api.c
@@ -320,6 +320,7 @@ static int ctp_sm_init(cci_plugin_ctp_t *plugin, uint32_t abi_ver, uint32_t flag
 		if (ret)
 			goto out;
 
+		sdev->pid = pid;
 		sdev->id = 0;
 
 		device->up = 1;
@@ -1332,6 +1333,23 @@ sm_free_conn(cci__conn_t *conn)
 		return;
 
 	if (sconn) {
+		int len = sizeof(*sconn->tx);
+
+		if (sconn->fifo != 0)
+			close(sconn->fifo);
+
+		if (sconn->mmap)
+			munmap(sconn->mmap, len);
+		if (sconn->peer_mmap)
+			munmap(sconn->peer_mmap, len);
+
+		len = sizeof(*sconn->rma);
+
+		if (sconn->rma_mmap)
+			munmap(sconn->rma_mmap, len);
+		if (sconn->peer_rma_mmap)
+			munmap(sconn->peer_rma_mmap, len);
+
 		free(sconn->name);
 		if (sconn->id != -1) {
 			ret = pthread_rwlock_wrlock(&sep->conns_lock);
@@ -1356,6 +1374,9 @@ sm_free_conn(cci__conn_t *conn)
 			}
 			sm_put_conn_id(sconn);
 		}
+		if (sconn->params)
+			free(sconn->params->data_ptr);
+		free(sconn->params);
 		free(sconn->rxs);
 		free(sconn->txs);
 		free(sconn);
@@ -1832,13 +1853,8 @@ static int ctp_sm_connect(cci_endpoint_t * endpoint, const char *server_uri,
 	debug(CCI_DB_CONN, "%s: connecting to %s", __func__, server_uri);
 
     out:
-	if (ret) {
-		if (params) {
-			free(params->data_ptr);
-			free(params);
-		}
+	if (ret)
 		sm_free_conn(conn);
-	}
 
 	CCI_EXIT;
 	return ret;
@@ -2036,6 +2052,7 @@ sm_handle_connect_reply(cci__ep_t *ep, void *buffer)
 
 	free(params->data_ptr);
 	free(params);
+	sconn->params = NULL;
 
 	evt->event.type = CCI_EVENT_CONNECT;
 	evt->event.connect.context = conn->connection.context;
-- 
2.5.5

